### PR TITLE
Input Handling in GVRf

### PR DESCRIPTION
**Pull request by rahul27**
This patch lays out the groundwork for input handling
in GVRf.

The introduced GVRInputManager allows input events
to be dispatched to through the scene graph. This enables
the GVRSceneObjects to implement user interaction using the
SensorEvents generated from GVRBaseSensors.

The framework for now supports Mice and Gamepads.

External Input devices can be added using the
GVRCursorController class.

GearVRf-DCO-1.0-Signed-off-by: Rahul Rudradevan
rahuldevan@gmail.com

### CODE REVIEW

**Comment by rahul27** [comment]
@liaxim @thomasflynn this pull request is ready for review. I will push the sample to the demos repo once this is through.

File: GVRf/Framework/src/org/gearvrf/GVRSceneObject.java [hunk 1]
```diff
@@ -704,6 +705,53 @@ public GVRSceneObject getChildByIndex(int index) {
...
+        return mSensor;
+    }
+
+    /**
+     * Remove the {@link GVRBaseSensor} from the object.
+     * 
+     * @return <code>true</code> if the sensor is successfully removed from the
+     *         object; <code>false</code> otherwise.
+     */
+    public boolean resetSensor() {
```
Comment (liaxim) [inline GVRf/Framework/src/org/gearvrf/GVRSceneObject.java]: We already have setSensor; this very thing could be achieved by setSensor(null) and people will have to deal with one api less. And it is pretty common "pattern".

**Comment by liaxim** [comment]
On a first glance this looks ready. I think the sample pull request needs to be created now so people can try it out.

File: GVRf/Framework/src/org/gearvrf/GVRSceneObject.java (hunk 1 above)
Comment (rahul27) [inline GVRf/Framework/src/org/gearvrf/GVRSceneObject.java]: Fixed this.

**Comment by rahul27** [comment]
@liaxim Added the input sample here: https://github.com/gearvrf/GearVRf-Demos/pull/3.

The demo makes use of the GVRTextViewSceneObject which has been deprecated. But I can't find a sample that makes use of the new GVRTextView class. 

Would it be okay if I move to the new GVRTextView in a separate patch once I figure out how to use it?

**Comment by liaxim** [comment]
Sure, no problem.

**Comment by liaxim** [comment]
Fyi, GVRTextView usage:

- create a GVRTextView
```
        mTextView = new GVRTextView(this);
        mTextView.layout(0, 0, 500, 500);
        mTextView.measure(500, 500);
        mTextView.setBackgroundColor(Color.WHITE);
        mTextView.setTextColor(Color.BLACK);
        mTextView.setText("UI TextView");
```

- create a GVRViewSceneObject
```
        mViewSceneObject = new GVRViewSceneObject(gvrContext, mTextView,
                 <a_mesh>);
```

- add the scene object to the scene

**Comment by liaxim** [comment]
Looks good to me.

In a (not so) future commit it would be desirable to destroy the controller's thread when a device is removed.

**Comment by liaxim** [comment]
For other reviewers: the sample is intentionally more complicated; it uses depth in the event resolution thus the use of isColliding(). This is not the expected general use-case so onSensorEvent should be more straightforward usually.

**Comment by rahul27** [comment]
@liaxim Sure I can stop the thread when the device is removed. Will add a patch to do the same.

I also plan to make another simple sample that demonstrates a simple use case, without isColliding().

**Comment by thomasflynn** [comment]
@rahul27 are you planning on another patch (for stopping the thread)?

**Comment by rahul27** [comment]
@thomasflynn will do it in a separate pull request.

**Pull request merged**